### ISSUE TITLE
No such device error

### ISSUE THREAD

**Issue by semperrin** [Microsemi/switchtec-kernel#114]
I am trying to transport data using IPoPCI via ntb_netdev. I am following the general outline given in step 4 of the "Non-Crosslink NTB connection for Linux" section:
https://docs.nvidia.com/drive/drive_os_5.1.6.1L/nvvib_docs/index.html#page/DRIVE_OS_Linux_SDK_Development_Guide/System%20Programming/sys_components_non_transparent_bridging.html
One machine is running Ubuntu 18.04 and has a PM8534 and the other machine is running Ubuntu 16.04 and has a PM40036 switch.

I am trying to load the kernels doing the following    

    modprobe ntb
    modprobe switchtec
    modprobe ntb_hw_switchtec
    modprobe ntb_transport
    modprobe ntb_netdev

When I load `ntb_transport` I get the first line:

    [  559.033375] Software Queue-Pair Transport over NTB, version 4
   
However, I do not get the second line:

    [  559.034097] switchtec switchtec0: ntb link up

Then if I try to load `ntb_netdev` I get the following:

    modprobe: ERROR: could not insert 'ntb_netdev': No such device
    
Can you provide me with any information on what may be causing this error?

**Comment by lsgunth** [comment]
Hmmm, it sounds like there may not be an NTB device. Could be a configuration issue with one of the switches or an number of other issues. Do you see anything in dmesg related to switchtec?

**Comment by semperrin** [comment]
After the command:

    sudo modprobe switchtec

I get the following in dmesg:

    [Feb23 14:51] switchtec: loaded.

And after the command:

    sudo modprobe ntb_transport

I get the following in dmesg:

    [ +31.978234] Software Queue-Pair Transport over NTB, version 4
 
 Those are the only messages I get in dmesg

**Comment by semperrin** [comment]
My issue seems quite similar to https://github.com/Microsemi/switchtec-kernel/issues/106 as when I also run switchtec-user command switchtec -list it returns:

    free(): invalid pointer
    Aborted (core dumped)

**Comment by semperrin** [comment]
The core dumping appears to be addressed in this pull request: https://github.com/Microsemi/switchtec-user/pull/263

**Comment by semperrin** [comment]
The PM40036 chip is within a Dolphin MXH930 NTB host adapter. Is the switch configuration you are referring to a HW or SW change? If SW, do you know how this configuration is managed and if it can be changed?

**Comment by lsgunth** [comment]
Yes, it's part of the firmware download. It's usually done with the Chiplink software. You might need to contact Microchip or your vendor (Dolphin) to get that setup. It's odd to me that a card designed for NTB doesn't have it configured correctly to begin with.

**Comment by semperrin** [comment]
Does the following output confirm that the switch is configured to be a NTB endpoint? My initial assumption was that it did. 

    $ lspci | grep Sierra
    07:00.0 PCI bridge: PMC-Sierra Inc. Device 4036
    07:00.1 Bridge: PMC-Sierra Inc. Device 4036
    07:00.2 System peripheral: PMC-Sierra Inc. Device 4036
    08:00.0 PCI bridge: PMC-Sierra Inc. Device 4036

If the above does not confirm that the switch is configured to be a NTB endpoint would: 

    ls /dev/switchtec*
listing a device confirm this? (currently that command lists no such devices for me) Is there some other way to confirm that the switch is or is not configured to be a NTB endpoint?

**Comment by Kendidi** [comment]
I tried to load switchtec drivers (ntb.ko, switchtec.ko and ntb_hw_switchtec.ko) for Dolphin MXP930.  But it appeared failing to find the `crosslink` partition while enumerating the BARs.

`[  673.144040] switchtec: loading out-of-tree module taints kernel.`
`[  673.144084] switchtec: module verification failed: signature and/or required key missing - tainting kernel`
`[  673.144836] switchtec 0000:01:00.1: enabling device (0000 -> 0002)`
`[  673.149496] switchtec switchtec0: Management device registered.`
`[  673.150487] switchtec: loaded.`
`[  673.391363] switchtec switchtec0: failed to register ntb device: -12`
`[ 1225.316886] switchtec switchtec0: unregistered.`

Is there anything needed to be configured on the adapter (e.g. update firmware) or modify the switchtec drivers before I can load the drivers successfully?  Please advise!

**Comment by lsgunth** [comment]
@semperrin The lspci trace doesn't tell us much. If you aren't getting a /dev/switchtec device then it's not configured for NTB and it is not configured with a management endpoint and there's not much you can do about that besides reconfigure it.

@Kendid You got a -12 error which is ENOMEM. This is not a likely error. It can happen if your system has no memory (unlikely) but it can also happen if the kernel is unable to map parts of the PCI bar. My guess is the switch's BARs are not configured appropriately for the driver and it's trying to map a bar that doesn't exist.

**Comment by Kendidi** [comment]
My card is Dolphin MXH930.  I wonder typically does anything need to be done to it before switchtec drivers can load, detect and register the device to the ntb core successfully?  Thanks.

**Comment by lsgunth** [comment]
No idea. I have no clue what that card is or how it's setup. You should probably contact Dolphin for support.

**Comment by Kendidi** [comment]
I see.

I wonder where can I learn more on how Microsemi/switchtec Crosslink works  and its requirements , etc?

Regarding the following in _**switchtec_ntb_init_crosslink()**_:
```
	if (bar_cnt < sndev->nr_direct_mw + 1) {
		dev_err(&sndev->stdev->dev,
			"Error enumerating crosslink partition\n");
		return -EINVAL;
	}
```

My current environment reports:
```
  bar_cnt = 1
  nr_direct_mw =3
```
What approach would you recommend to debug thist?  Why ` bar_cnt` should be >= `nr_direct_mw + 1` in order for it to proceed?  Thanks!

**Comment by Kendidi** [comment]
It appears crosslink config is looking for the following BAR addresses from vEP, am I correct?
```
1.  0x00_0000_0000
2.  0x10_0000_0000
3.  0x20_0000_0000
```

**Comment by Kendidi** [comment]
I see.  Thank you Logan for the confirmation and info,!

So if server A wants to share a NVMe SSD to server B, can both NTB adapters on server A and server B have the same configuration/firmware?  Or should they be different since the server's role are different?  Thanks.

**Comment by lsgunth** [comment]
I'm not aware of any solutions for NVMe sharing that are not complicated and proprietary.  So you're pretty much on your own if you want to implement something like that. 

The point of cross link was to make both machines symmetric. (So cases where each machine has a switch which connect to each other). In these cases the configuration for each switch should be the identical.

Just as it's possible to write a driver to share the NVMe drive between two partitions on a single switch, it should, at least theoretically, be possible to share an NVMe drive over a cross link setup. Though, I don't know if anyone has actually ever tried this.

**Comment by Kendidi** [comment]
Okie.  Thanks Logan!

I tried to look up for more info. on Microsemi ChipLink but I couldn't find anything about it.  It is rebranded to another product or something?

**Comment by jborz27** [comment]
Hi Kendidi,

Microsemi was purchased by Microchip.  You can find info on ChipLink on Microchip’s YouTube channel and on their web site.  Some resources will be restricted to customers.

**Comment by Kendidi** [comment]
Thanks @jborz27 !  I will see where it can be downloaded.

**Comment by Kendidi** [comment]
Any suggestions on where to and how to get Microsemi/Microchip Chiplink?

**Comment by krlc** [comment]
@Kendidi  Did you end up getting ChipLink? If so, how?
